tiledb/sm/filesystem/vfs.cc

```diff
@@ -381,18 +382,108 @@ void VFS::move_dir(const URI& old_uri, const URI& new_uri) const {
+  auto &src_fs{get_fs(src)}, &dest_fs{get_fs(dest)};
+
+  // Deque which stores the buffers passed between the readers and writer.
+  ProducerConsumerQueue<
+      std::variant<tdb_shared_ptr<std::vector<char>>, std::exception_ptr>>
+      buffer_queue;
+
+  // By default, the buffer size is 10 MB, unless the filesize is smaller.
+  uint64_t filesize = src_fs.file_size(src);
+  uint64_t buffer_size = std::min(filesize, (uint64_t)10485760);  // 10 MB
+
+  // The maximum number of buffers the reader may allocate.
+  const int max_buffer_count = 10;
+
+  // Atomic counter of the buffers allocated by the reader.
+  // May not exceed `max_buffer_count`.
+  std::atomic<int> buffer_count = 0;
+
+  // Flag indicating an ongoing read. The reader will stop once set to `false`.
+  std::atomic<bool> reading = true;
+
+  // Reader
+  ThreadPool::Task read_task = io_tp_->execute([&] {
+    uint64_t offset = 0;
+    while (reading) {
+      tdb_unique_ptr<std::vector<char>> buffer(
+          tdb_new(std::vector<char>, buffer_size));
+      try {
+        uint64_t bytes_read =
+            src_fs.read(src, offset, buffer->data(), buffer->size());
+        if (bytes_read > 0) {
+          buffer->resize(bytes_read);
+          buffer_queue.push(std::move(buffer));
+          offset += bytes_read;
+        }
+
+        // Once the read is complete, drain the queue and exit the reader.
+        // Note: drain() shuts down the queue without removing elements.
+        // The write fiber will be notified and write the remaining chunks.
+        if (offset >= filesize) {
+          buffer_queue.drain();
+          reading = false;
+          break;
+        }
+      } catch (...) {
+        // Enqueue caught-exceptions to be handled by the writer.
+        buffer_queue.push(std::current_exception());
+      }
+      buffer_count++;
+
+      io_tp_->wait_until([&]() { return buffer_count < max_buffer_count; });
+    }
+    return Status::Ok();
+  });
+
+  // Writer
+  uint64_t bytes_written = 0;
+  while (true) {
+    // Allow the ProducerConsumerQueue to wait for an element to be enqueued.
+    auto buffer_queue_element = buffer_queue.pop_back();
+    if (!buffer_queue_element.has_value()) {
+      // Stop writing once the queue is empty (reader finished reading file).
+      break;
+    }
+
+    auto& buffer = buffer_queue_element.value();
+    // Rethrow exceptions enqueued by read.
+    if (std::holds_alternative<std::exception_ptr>(buffer)) {
+      std::rethrow_exception(std::get<std::exception_ptr>(buffer));
+    }
+
+    auto& writebuf = std::get<0>(buffer);
+    try {
+      dest_fs.write(dest, writebuf->data(), writebuf->size());
+      buffer_count--;
+    } catch (...) {
+      reading = false;  // Stop the reader.
+      throw;
+    }
+    bytes_written += writebuf->size();
+  }
+
+  if (bytes_written != filesize) {
+    throw VFSException("Failed to copy file due to internal error.");
+  }
+
```

Comment:
Under which circumstances is this condition expected to hold? The read loop is not going to stop until it reads all bytes. Should we remove it?